x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MlDailyMaintenanceService.java

```diff
@@ -252,6 +292,159 @@ private void triggerNlpMaintenance() {
+
+                // Make sure we use a fresh IndicesAliasesRequestBuilder, the original one may have changed internal state.
+                IndicesAliasesRequestBuilder localAliasRequestBuilder = originSettingClient.admin()
+                    .indices()
+                    .prepareAliases(
+                        MachineLearning.HARD_CODED_MACHINE_LEARNING_MASTER_NODE_TIMEOUT,
+                        MachineLearning.HARD_CODED_MACHINE_LEARNING_MASTER_NODE_TIMEOUT
+                    );
+
+                // Execute the cleanup, no need to propagate the original failure.
+                removeRolloverAlias(indexName, rolloverAlias, localAliasRequestBuilder, listener);
+            } else {
+                listener.onFailure(e);
+            }
+        });
+
+        // 3 Update aliases
+        ActionListener<String> rolloverListener = ActionListener.wrap(newIndexNameResponse -> {
+            MlIndexAndAlias.addIndexAliasesRequests(aliasRequestBuilder, index, newIndexNameResponse, clusterState);
+            // On success, the rollover alias may have been moved to the new index, so we attempt to remove it from there.
+            // Note that the rollover request is considered "successful" even if it didn't occur due to a condition not being met
+            // (no exception will be thrown). In which case the attempt to remove the alias here will fail with an
+            // IndexNotFoundException. We handle this case with a secondary listener.
+            removeRolloverAlias(newIndexNameResponse, rolloverAlias, aliasRequestBuilder, aliasListener);
+        }, e -> {
+            // If rollover fails, we must still clean up the temporary alias from the original index.
+            // The index name is either the original one provided or the original with a suffix appended.
+            var indexName = MlIndexAndAlias.has6DigitSuffix(index) ? index : index + MlIndexAndAlias.FIRST_INDEX_SIX_DIGIT_SUFFIX;
+            // Execute the cleanup, no need to propagate the original failure.
+            removeRolloverAlias(indexName, rolloverAlias, aliasRequestBuilder, aliasListener);
+        });
+
+        // 2 rollover the index alias to the new index name
+        ActionListener<IndicesAliasesResponse> getIndicesAliasesListener = ActionListener.wrap(getIndicesAliasesResponse -> {
+            rollover(originSettingClient, rolloverAlias, newIndexName, rolloverListener);
+        }, rolloverListener::onFailure);
+
+        // 1. Create necessary aliases
+        MlIndexAndAlias.createAliasForRollover(originSettingClient, index, rolloverAlias, getIndicesAliasesListener);
+    }
+
+    // public for testing
+    public void setRolloverConditions(RolloverConditions rolloverConditions) {
+        this.rolloverConditions = Objects.requireNonNull(rolloverConditions);
+    }
+
+    // public for testing
+    public void triggerRollResultsIndicesIfNecessaryTask(ActionListener<AcknowledgedResponse> finalListener) {
```

Comment:
The method mixes index discovery, filtering, rollover execution, and error collection. It would improve readability if you would extract focused methods like `findIndicesNeedingRollover()`, `rolloverIndexSafely()`, `handleRolloverResults()`.